File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	long long k, n;
	cin >> k >> n;

	long long lan[k];
	long long max = -1;
	for (int i = 0; i < k; i++) {
		cin >> lan[i];
		if (lan[i] > max) max = lan[i];
	}

	long long start = 1;
	long long mid;
	long long end = max;
	long long cnt;
	long long answer = 0;
	while (start <= end) {
		cnt = 0;
		mid = (start+end)/2;
		for (int i = 0; i < k; i++) {
			cnt += lan[i] / mid;
		}

		if (cnt >= n) {
			answer = mid;
			start = mid + 1;
		}
		else {
			end = mid - 1;
		}
	} 
	cout << answer << '\n';
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

k, n = map(int, sys.stdin.readline().split())

lst = []

for i in range(k):
    tmp = int(sys.stdin.readline().rstrip())
    lst.append(tmp)

left = 1
right = max(lst)
mid = (left+right)//2

answer = 0

while left <= right:
    mid = (left+right)//2

    cnt = 0

    for length in lst:
        cnt += length//mid

    if cnt >= n:
        answer = mid
        left = mid + 1

    else:
        right = mid - 1

    
print(answer)